File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/eal_to_al.py
# Ham chuyen tu danh sach ke mo rong ve danh sach ke
def eal_to_al(eal, n):
    al = [[] for _ in range(n)]  # Khoi tao danh sach ke rong

    for i in range(n):
        for (j, _) in eal[i]:  # Bo qua trong so, chi lay dinh j
            al[i].append(j)

    return al

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
eal = []

print("Nhap danh sach ke mo rong:")
for i in range(n):
    line = input(f"Dinh {i} (dinh,trongso cach nhau boi space): ")
    pairs = []
    if line.strip() != "":
        tokens = line.split()
        for token in tokens:
            u, w = map(int, token.strip("()").split(","))
            pairs.append((u, w))
    eal.append(pairs)

# Chuyen doi va in ra danh sach ke
al = eal_to_al(eal, n)

print("Danh sach ke (AL):")
for i, neighbors in enumerate(al):
    print(f"{i}: {neighbors}")

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/eal_to_al.cpp
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

// Ham chuyen tu danh sach ke mo rong ve danh sach ke
vector<vector<int>> ealToAl(const vector<vector<pair<int, int>>>& eal, int n) {
    vector<vector<int>> al(n);  // Khoi tao danh sach ke

    for (int i = 0; i < n; ++i) {
        for (const auto& [j, _] : eal[i]) {
            al[i].push_back(j);  // Chi lay dinh j, bo qua trong so
        }
    }

    return al;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    vector<vector<pair<int, int>>> eal(n);
    cout << "Nhap danh sach ke mo rong:\n";
    for (int i = 0; i < n; ++i) {
        cout << "Dinh " << i << " (dinh,trongso): ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        string token;
        while (ss >> token) {
            int u, w;
            sscanf(token.c_str(), "(%d,%d)", &u, &w);
            eal[i].emplace_back(u, w);
        }
    }

    // Chuyen doi va in danh sach ke
    auto al = ealToAl(eal, n);

    cout << "Danh sach ke (AL):\n";
    for (int i = 0; i < n; ++i) {
        cout << i << ": ";
        for (int v : al[i]) {
            cout << v << " ";
        }
        cout << endl;
    }

    return 0;
}